pilot/pkg/model/push_context.go

```diff
@@ -1839,16 +1839,64 @@ func (ps *PushContext) initSidecarScopes(env *Environment) {
 	// Root namespace can have only one sidecar config object
 	// Currently we expect that it has no workloadSelectors
 	var rootNSConfig *config.Config
-	ps.sidecarIndex.sidecarsByNamespace = make(map[string][]*SidecarScope, len(sidecarConfigs))
 	for i, sidecarConfig := range sidecarConfigs {
-		ps.sidecarIndex.sidecarsByNamespace[sidecarConfig.Namespace] = append(ps.sidecarIndex.sidecarsByNamespace[sidecarConfig.Namespace],
-			ConvertToSidecarScope(ps, &sidecarConfig, sidecarConfig.Namespace))
-		if rootNSConfig == nil && sidecarConfig.Namespace == ps.Mesh.RootNamespace &&
+		if sidecarConfig.Namespace == ps.Mesh.RootNamespace &&
 			sidecarConfig.Spec.(*networking.Sidecar).WorkloadSelector == nil {
 			rootNSConfig = &sidecarConfigs[i]
+			break
 		}
 	}
 	ps.sidecarIndex.meshRootSidecarConfig = rootNSConfig
+
+	ps.sidecarIndex.sidecarsByNamespace = make(map[string][]*SidecarScope)
+	if features.ConvertSidecarScopeConcurrency > 1 {
+		ps.concurrentConvertToSidecarScope(sidecarConfigs)
+	} else {
+		for _, sidecarConfig := range sidecarConfigs {
+			ps.sidecarIndex.sidecarsByNamespace[sidecarConfig.Namespace] = append(ps.sidecarIndex.sidecarsByNamespace[sidecarConfig.Namespace],
+				ConvertToSidecarScope(ps, &sidecarConfig, sidecarConfig.Namespace))
+		}
+	}
+}
+
+func (ps *PushContext) concurrentConvertToSidecarScope(sidecarConfigs []config.Config) {
+	if len(sidecarConfigs) == 0 {
+		return
+	}
+	type taskItem struct {
+		idx int
+		cfg config.Config
+	}
+
+	var wg sync.WaitGroup
+	taskItems := make(chan taskItem)
+	// note: sidecarConfigs order matters
+	sidecarScopes := make([]*SidecarScope, len(sidecarConfigs))
+	for i := 0; i < features.ConvertSidecarScopeConcurrency; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for {
+				item, ok := <-taskItems
+				if !ok {
+					break
+				}
+				sc := ConvertToSidecarScope(ps, &item.cfg, item.cfg.Namespace)
+				sidecarScopes[item.idx] = sc
```

Comment:
Missing lock here

Reply:
Since the space of `sidecarScopes` is allocated in advance and we access the index of the slice, there is no data race here.